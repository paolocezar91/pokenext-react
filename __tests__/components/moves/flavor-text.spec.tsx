import FlavorText from "@/components/moves/flavor-text";
import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import { IMove } from "pokeapi-typescript";

jest.mock("@heroicons/react/24/solid", () => ({
  ChevronLeftIcon: () => <div>LeftIcon</div>,
  ChevronRightIcon: () => <div>RightIcon</div>,
}));

jest.mock("next-intl", () => ({
  useTranslations: () => (key: string) => key,
}));

jest.mock("@/context/user-context", () => ({
  useUser: () => ({
    settings: { descriptionLang: "en" },
  }),
}));

describe("FlavorText Component", () => {
  const mockMoveData = {
    name: "fire-blast",
    flavor_text_entries: [
      {
        flavor_text: "First flavor text",
        language: { name: "en" },
        version_group: { name: "red-blue" },
      },
      {
        flavor_text: "Second flavor text",
        language: { name: "en" },
        version_group: { name: "gold-silver" },
      },
      {
        flavor_text: "Non-English text",
        language: { name: "jp" },
        version_group: { name: "diamond-pearl" },
      },
      {
        flavor_text: "Third flavor text",
        language: { name: "en" },
        version_group: { name: "sword-shield" },
      },
    ],
  } as unknown as IMove;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render the component with initial state", () => {
    render(<FlavorText moveData={mockMoveData} />);

    // Test title transformation
    expect(screen.getByRole("heading")).toHaveTextContent(
      "moves.flavorText.title",
    );

    // Test initial flavor text
    expect(screen.getByText("First flavor text")).toBeInTheDocument();

    // Test version group normalization
    expect(screen.getByText("(Red/Blue)")).toBeInTheDocument();

    // Test pagination display
    expect(screen.getByText("1 / 3")).toBeInTheDocument();
  });

  it("should filter non-English flavor texts", () => {
    render(<FlavorText moveData={mockMoveData} />);

    // Only English texts should be shown (3 out of 4)
    expect(screen.queryByText("Non-English text")).not.toBeInTheDocument();
    expect(screen.getByText("1 / 3")).toBeInTheDocument();
  });

  it("should navigate between flavor texts", () => {
    render(<FlavorText moveData={mockMoveData} />);

    const prevButton = screen.getByTestId("previous-button");
    const nextButton = screen.getByTestId("next-button");

    // Initial state - prev button disabled
    expect(prevButton).toBeDisabled();
    expect(nextButton).not.toBeDisabled();

    // Go to next flavor text
    fireEvent.click(nextButton);
    expect(screen.getByText("Second flavor text")).toBeInTheDocument();
    expect(screen.getByText("(Gold/Silver)")).toBeInTheDocument();
    expect(screen.getByText("2 / 3")).toBeInTheDocument();
    expect(prevButton).not.toBeDisabled();

    // Go to next again
    fireEvent.click(nextButton);
    expect(screen.getByText("Third flavor text")).toBeInTheDocument();
    expect(screen.getByText("(Sword/Shield)")).toBeInTheDocument();
    expect(screen.getByText("3 / 3")).toBeInTheDocument();
    expect(nextButton).toBeDisabled();

    // Go back to previous
    fireEvent.click(prevButton);
    expect(screen.getByText("Second flavor text")).toBeInTheDocument();
    expect(screen.getByText("2 / 3")).toBeInTheDocument();
  });

  it("should handle single flavor text case", () => {
    const singleFlavorData = {
      ...mockMoveData,
      flavor_text_entries: [mockMoveData.flavor_text_entries[0]],
    };

    render(<FlavorText moveData={singleFlavorData} />);

    const prevButton = screen.getByTestId("previous-button");
    const nextButton = screen.getByTestId("next-button");

    expect(prevButton).toBeDisabled();
    expect(nextButton).toBeDisabled();
    expect(screen.getByText("1 / 1")).toBeInTheDocument();
  });

  it("should match snapshot", () => {
    const singleFlavorData = {
      ...mockMoveData,
      flavor_text_entries: [mockMoveData.flavor_text_entries[0]],
    };
    const { asFragment } = render(<FlavorText moveData={singleFlavorData} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
