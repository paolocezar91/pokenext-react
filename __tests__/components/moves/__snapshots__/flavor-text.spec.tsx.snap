// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlavorText Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flavor w-full mt-2"
  >
    <h3
      class="w-fit text-lg mb-4"
    >
      moves.flavorText.title
    </h3>
    <div
      class="flex items-end relative items-start justify-between pb-6"
    >
      <p>
        First flavor text
      </p>
      <div
        class="flex"
      >
        <button
          aria-label="actions.previous flavor text"
          class="text-xs mr-1 h-5 rounded hover:bg-(--pokedex-red) disabled:bg-gray-600 transition-colors btn btn-primary"
          data-testid="previous-button"
          disabled=""
          type="button"
        >
          <div>
            LeftIcon
          </div>
        </button>
        <button
          aria-label="actions.next flavor text"
          class="text-xs ml-1 h-5 rounded hover:bg-(--pokedex-red) disabled:bg-gray-600 transition-colors btn btn-primary"
          data-testid="next-button"
          disabled=""
          type="button"
        >
          <div>
            RightIcon
          </div>
        </button>
      </div>
      <div
        class="absolute bottom-[-8px] right-0 border-t-1 border-t-white text-sm"
      >
        <small
          class="mr-2"
        >
          (Red/Blue)
        </small>
        1 / 1
      </div>
    </div>
  </div>
</DocumentFragment>
`;
